### PR TITLE
[4.0] spinner as web component

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue # .

### Summary of Changes
- Drops the code from the core.js for the spinner (already deprecated in j3) and introduces a web component
- Some minor changes in the build module for the web components. The settings.json will now allow custom paths in the media folder. So someone could do some PRs to move the fields custom elements to their appropriate place (/media/system/js/fields/)
- Seperated the logic from the mark up for the category field (someone really needs to do this for all the mark up that is hidden in fields or some helpers!!!)
- Reworked the custom fields part that injected some inline jQuery script. So now the script is part of the category field (that is the right place for the js to live) and the helper just passes some data to the field. The js is still inline so someone should just move it to a file...
- Patched the custom fields to work with the new element
- Patched the installation to work with the new element
- Patched the com_associations to work with the new element

### Testing Instructions
Install Joomla, spinner is ok
Make sure that you have custom fields assigned to one category. Edit an article and change its cat to the one that custom fields are attached. spinner is ok
Try a multilingual and check if the spinner works ok

### Documentation Changes Required

In 4.0 in each page that a spinner is needed the following steps will need to be done:

- Include the web component
```php
HTMLHelper::_('webcomponent', 'system/webcomponents/joomla-core-loader.min.js', ['relative' => true, 'version' => 'auto']);
```

- To show the spinner all we need to do is create an element:
```js
spinner = document.createElement('joomla-core-loader');

// Assuming that the spinner is full screen
document.body.appendChild(spinner);

/*
if we need to append the spinner to some other container
all we have to do is replace document.body with the element 
that we want the spinner to appear
*/
```

- To remove the spinner all we need to do is destroy the element:
```js
spinner = document.querySelector('joomla-core-loader');

spinner.parentNode.removeChild(spinner);
```

 if you want to use the same codebase for different versions (j3 and j4) here is a quick code:

```js
if (Joomla.loadingLayer && typeof Joomla.loadingLayer === 'function') {
 // We are in J3 so use the old method
 Joomla.loadingLayer('show');
} else {
 // We are in the future
spinner = document.createElement('joomla-core-loader');
document.body.appendChild(spinner);
}
```

### CODE REVIEW

File: administrator/components/com_fields/Helper/FieldsHelper.php [hunk 1]
```diff
@@ -332,24 +332,11 @@ public static function prepareForm($context, Form $form, $data)
...
-				jQuery('input[name=task]').val('" . $section . ".reload');
-				element.form.submit();
-			}
-			jQuery( document ).ready(function() {
-				Joomla.loadingLayer('load');
-				var formControl = '#" . $form->getFormControl() . "_catid';
-				if (!jQuery(formControl).val() != '" . $assignedCatids . "'){jQuery(formControl).val('" . $assignedCatids . "');}
-			});"
-			);
+			 */
```
Comment (laoneo) [inline administrator/components/com_fields/Helper/FieldsHelper.php:334]: Needs to get removed or?

File: layouts/joomla/form/field/category.php [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+	$html[] = HTMLHelper::_('select.genericlist', $options, $name, trim($attr), 'value', 'text', $value, $id);
+}
+
+echo implode($html);
+
+// Preload spindle-wheel when we need to submit form due to category selector changed
+\Joomla\CMS\Factory::getDocument()->addScriptDeclaration(
+<<<JS
+document.addEventListener('DOMContentLoaded', function() {
```
Comment (laoneo) [inline layouts/joomla/form/field/category.php]: This part should be added only when custom fields are involved

File: build/media/com_associations/js/sidebyside.js [hunk 3]
```diff
@@ -113,7 +110,7 @@ jQuery(document).ready(function($) {
 			target.setAttribute('data-language', selected.split(':')[0]);
 
 			// Iframe load start, show Joomla loading layer.
-			Joomla.loadingLayer('show');
+            document.body.appendChild(document.createElement('joomla-core-loader'));
```
Comment (brianteeman) [inline build/media/com_associations/js/sidebyside.js:113]: indent

File: build/media/com_associations/js/sidebyside.js [hunk 4]
```diff
@@ -168,7 +165,8 @@ jQuery(document).ready(function($) {
 		});
 
 		// Iframe load finished, hide Joomla loading layer.
-		Joomla.loadingLayer('hide');
+        var spinner = document.querySelector('joomla-core-loader');
```
Comment (brianteeman) [inline build/media/com_associations/js/sidebyside.js:168]: indent

File: build/media/webcomponents/js/core-loader/core-loader.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    /**
+     * Creates a custom element with the default spinner of the Joomla logo
+     */
+    class JoomlaCoreLoader extends HTMLElement {
+        constructor() {
+            super();
+
+            const template = document.createElement('template');
+            template.innerHTML = `<style>{{CSS_CONTENTS_PLACEHOLDER}}</style>
+<div><span class="yellow"></span><span class="red"></span><span class="blue"></span><span class="green"></span><p>&reg;</p></div>`;
```
Comment (brianteeman) [inline build/media/webcomponents/js/core-loader/core-loader.js:13]: indent

File: installation/template/js/setup.js [hunk 6]
```diff
@@ -12,7 +12,7 @@
 Joomla.setlanguage = function(form) {
 	var data = Joomla.serialiseForm(form);
 
-	Joomla.loadingLayer("show");
+    document.body.appendChild(document.createElement('joomla-core-loader'));
```
Comment (brianteeman) [inline installation/template/js/setup.js]: indent

File: installation/template/js/setup.js [hunk 7]
```diff
@@ -30,14 +30,17 @@ Joomla.setlanguage = function(form) {
 
 			if (response.error) {
 				Joomla.renderMessages({'error': [response.message]});
-				Joomla.loadingLayer("hide");
+				var el = document.querySelector('joomla-core-loader')
+				el.parentNode.removeChild(el);
 			} else {
-				Joomla.loadingLayer("hide");
+                var el = document.querySelector('joomla-core-loader')
```
Comment (brianteeman) [inline installation/template/js/setup.js:36]: indent

**Comment by brianteeman** [comment]
There are multiple issues being addressed in this one pr and it really should be one issue per pr. It makes testing much more reliable and avoids parts of the pr not being tested etc

**Comment by dgrammatiko** [comment]
It’s a take it or leave it...

File: administrator/components/com_fields/Helper/FieldsHelper.php (hunk 1 above)
Comment (dgrammatiko) [reply administrator/components/com_fields/Helper/FieldsHelper.php:334]: Yup this inline script cannot be overridden unless you create a plugin (which might not even possible to do in j4 due to name spacing). Asset loading should happen only in layouts

File: layouts/joomla/form/field/category.php (hunk 2 above)
Comment (dgrammatiko) [reply layouts/joomla/form/field/category.php]: It needs a wrapper similar to the one one top

**Comment by brianteeman** [comment]
Thats a shame as its not the right thing to do (as you well know) and its the reason why whe in the past pr's with multiple issues were merged we are wasting time undoing all the mistakes.

**Comment by Fedik** [comment]
maybe can you please skip the category field for now? it will have a huge conflict with #22263

File: build/media/com_associations/js/sidebyside.js (hunk 3 above)
Comment (dgrammatiko) [reply build/media/com_associations/js/sidebyside.js:113]: Actually this file needs to be converted to es6 and vanilla js. The indentation won’t solve the real problems here

**Comment by dgrammatiko** [comment]
@Fedik unfortuanatelly this is a really poor implementation with so many anti patterns so at some point it needs to be done.
- field markup is not overridable
- field specific js is loaded in a helper file in com_fields

I know it’s a pain but at some point these needs to be done correctly...

File: build/media/webcomponents/js/core-loader/core-loader.js (hunk 5 above)
Comment (dgrammatiko) [reply build/media/webcomponents/js/core-loader/core-loader.js:13]: This is correctly indented, the contents are part of the template therefore they don’t obey the the cs of the js, something like `<<<JS
>>>` in php world

**Comment by brianteeman** [comment]
As I am currently wasting a lot of time redoing a lot of work that was committed in error because it was not tested properly because it was buried in a big pr with other things we must stop accepting any pr that addresses multiple issues at the same time. It is a waste of people's limited time to have to redo things that were not able to be tested!

**Comment by dgrammatiko** [comment]
Ok good luck

**Pull request closed without merging**